src/jsonpatcherproxy.js

```diff
@@ -150,176 +84,202 @@ const JSONPatcherProxy = (function() {
-              And we know that because we mark moved proxies with `inherited` flag when we move them
-              
-              it is a good idea to remove this flag if we come across it here, in deleteProperty trap.
-              We DO want to revoke the proxy if it was removed again.
-              */
+      const destinationPropKey = parentPath + '/' + escapePathComponent(key);
 
-              revokableProxyInstance.inherited = false;
-            } else {
-              instance.objectsPathsMap.delete(
-                revokableProxyInstance.originalObject
-              );
-              instance.disableTrapsForProxy(revokableProxyInstance);
-              instance.proxifiedObjectsMap.delete(target[key]);
-            }
-          }
+      instance.defaultCallback({
+        op: 'remove',
+        path: destinationPropKey
+      });
+      const revokableProxyInstance = instance.proxifiedObjectsMap.get(
+        target[key]
+      );
+
+      if (revokableProxyInstance) {
+        if (revokableProxyInstance.inherited) {
+          /*
+            this is an inherited proxy (an already proxified object that was moved around), 
+            we shouldn't revoke it, because even though it was removed from path1, it is indeed used in path2.
+            And we know that because we mark moved proxies with `inherited` flag when we move them
+
+            it is a good idea to remove this flag if we come across it here, in deleteProperty trap.
+            We DO want to revoke the proxy if it was removed again.
+          */
+          revokableProxyInstance.inherited = false;
+        } else {
+          instance.parenthoodMap.delete(revokableProxyInstance.originalObject);
+          instance.disableTrapsForProxy(revokableProxyInstance);
+          instance.proxifiedObjectsMap.delete(target[key]);
         }
-        const result = Reflect.deleteProperty(target, key);
-        /* we need the Reflection to occur before we map paths to their object */
-        instance._resetCachedProxiesPaths(instance.cachedProxy, '');
-        return result;
       }
+    }
+    return Reflect.deleteProperty(target, key);
+  }
+  /* pre-define resume and pause functions to enhance constructors performance */
+  function resume() {
+    this.defaultCallback = operation => {
+      this.isRecording && this.patches.push(operation);
+      this.userCallback && this.userCallback(operation);
+    };
+    this.isObserving = true;
+  }
+  function pause() {
+    this.defaultCallback = () => {};
+    this.isObserving = false;
+  }
+  /**
+    * Creates an instance of JSONPatcherProxy around your object of interest `root`. 
+    * @param {Object|Array} root - the object you want to wrap
+    * @param {Boolean} [showDetachedWarning] - whether to log a warning when a detached sub-object is modified @see {@link https://github.com/Palindrom/JSONPatcherProxy#detached-objects} 
```

Comment:
is `true` a default value? if so I'd use the `[showDetachedWarning=true]` syntax